tests/unit/test.watcher.js

```diff
@@ -207,3 +208,57 @@ describe('watcher', () => {
   });
 
 });
+
+describe('--watch-ignored is passed in', () => {
+  it('does not change if ignored file is touched', () =>
+    withTempDir(async (tmpDir) => {
```

Comment:
> Could I please also ask how was race condition happening earlier?

@ankushduacodes well, I have the feeling that the race is still there, the change we made were mainly aiming to reduce the intermittent failures chances as much as possible.

There may be multiple reasons for the race:
- one is the debounce time we do have (changes of making that trigger a failure reduce by making as able to customize the debounce time to a lower value during the test execution)

- watchpack's watch method takes a third parameter that is the startTime, which is then used to check if the file mtime is >= of the given startTime and not emit the change event if it is not (e.g. see [here](https://github.com/webpack/watchpack/blob/40ddaa9c895e83c5eb490abca6834ae1e71aa8cc/lib/DirectoryWatcher.js#L45-L49) and [here](https://github.com/webpack/watchpack/blob/40ddaa9c895e83c5eb490abca6834ae1e71aa8cc/lib/DirectoryWatcher.js#L224-L236) in the watchpack sources), we do set the startTime to Date.now() in how watcher utils.

- watchpak seems to be internally using a FS_ACCURACY variable that may effect the behavior (in other words if the change event will be emitted or not), initially defined [here](https://github.com/webpack/watchpack/blob/40ddaa9c895e83c5eb490abca6834ae1e71aa8cc/lib/DirectoryWatcher.js#L15).

- the underlying OS feature to watch for changes will be different for each OS and may present slightly different behavior and accuracy 

These are the once that I did notice when I looked into what may be triggering the unexpected failures we were looking into.

Yesterday I did tried a kind of stress test to double-check if the test was going to fail intermittently over a high number of runs, mocha doesn't seem to support any command line option to do that without changing the tests themselves and so I tweaked the test locally as follows:

```
it('does not change if ignored file is touched', async () => {
    for (var i = 0; i < 200; i++) {
      await withTempDir(async (tmpDir) => {
        ...
      });
    }
});
```

and then I execute it by launching mocha manually on this particular test case and with a longer test timeout set using the mocha cli options:

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 800000
```

This was stress test version of the test case was still failing for me from time to time, I would be ok to defer looking more into it and see if those intermittent failures are going to happen for real after we land it (and if it does fail intermittently then we can look into apply the necessary changes to reduce it a bit more as part of a follow up).

Reply:
@rpl Thank you for the detailed explanation of why race condition is occuring!..

You are right, the stress test is failing for me as well. Only the first iteration of the test loop is passing for me, Is that the case on your machine as well?

Reply:
> @rpl Thank you for the detailed explanation of why race condition is occurring!..

My pleasure, feel always free to ask us more context when you are curious or have doubts on some details (even they are not blocking issue and seems secondary details), in my opinion discussing about those kind of questions is an important part of a mentored bug (as much as getting the particular pull request done or good first bug fixed) and I'm absolutely happy to answer them with as much context I can. 

> You are right, the stress test is failing for me as well. Only the first iteration of the test loop is passing for me, Is that the case on your machine as well?

Interesting, on my machine it was able to run a bunch of times before triggering one intermittent failure.

The following changes were making the stress test a bit more stable on my machine, would you like to try if they do also on macos? that may be interesting to anticipate how bad the intermittency would be in practice.

- reduced the value of `debounceTime` (I did set it to `0`)
- changed `waitDebounce`'s setTimeout second parameter from `debounceTime * 2` to `10` (or a value a bit higher if based on the behavior of the stress test experiment it seems to make the test a bit more stable)
- added an additional `await waitDebounce();` between `await fs.writeFile(...)` and `sinon.assert.callCount(onChange, ...)` (to leave some time for the onChange spy to actually be called, I had the impression that it was another potential source of intermittency)

If you decide to give it a try, let me know if these tweaks did make the test a bit more stable on MacOS as they seem to do while running on Linux.

Reply:
@rpl I tried running tests multiple times with `debounceTime` set to zero and different `waitDebounce`'s `setTimeout`'s second param set to 10, 20, 30, 40, 100, 500, 1000.
Added waitDebounce between fs.writeFile and assert statement

<details open>

<summary>In case of 10:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 17:49:46


  watcher
    --watch-ignored is passed in
iteration 0
      1) does not change if ignored file is touched


  0 passing (1m)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     Error: Timeout of 80000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)
```

</details>
<details open>

<summary>In case of 20:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 17:56:42


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
      1) does not change if ignored file is touched


  0 passing (590ms)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)

```

</details>
<details open>

<summary>In case of 30:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 17:58:25


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
      1) does not change if ignored file is touched


  0 passing (593ms)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)
```

</details>

<details open>

<summary>In case of 40:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 18:03:25


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
Passed 1
iteration 2
      1) does not change if ignored file is touched


  0 passing (986ms)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)


```

</details>
<details open>

<summary>In case of 100:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 18:04:15


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
Passed 1
iteration 2
      1) does not change if ignored file is touched


  0 passing (2s)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)


```

</details>
<details open>

<summary>In case of 500:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 18:05:36


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
      1) does not change if ignored file is touched


  0 passing (5s)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)


```

</details>
<details open>

<summary>In case of 1000:</summary>
<br>

```
npx mocha -r @babel/register -r tests/setup.js tests/unit/test.watcher.js -f "ignored file" -t 80000           (base) 18:06:23


  watcher
    --watch-ignored is passed in
iteration 0
Passed 0
iteration 1
      1) does not change if ignored file is touched


  0 passing (9s)
  1 failing

  1) watcher
       --watch-ignored is passed in
         does not change if ignored file is touched:
     expected spy to be called 0 times but was called once
    spy() at onChange (/Users/ankushdua/Documents/GitHub/web-ext/node_modules/debounce/index.js:40:21)
  AssertError: expected spy to be called 0 times but was called once
      spy() at onChange (node_modules/debounce/index.js:40:21)
      at Object.fail (node_modules/sinon/lib/sinon/assert.js:110:25)
      at failAssertion (node_modules/sinon/lib/sinon/assert.js:67:20)
      at Object.callCount (node_modules/sinon/lib/sinon/assert.js:151:17)
      at assertOnChange (tests/unit/test.watcher.js:249:26)
      at /Users/ankushdua/Documents/GitHub/web-ext/tests/unit/test.watcher.js:253:11
      at Context.<anonymous> (tests/unit/test.watcher.js:214:9)


```

</details>


it seems to be highly inconsistent, I also tried running each of these tests above multiple times, and gave me very inconsistent results when it came to asserting the calls to `onChange`, a few of these timed out(Probably because of race condition).

### Doubts:
> one is the debounce time we do have (changes of making that trigger a failure reduce by making as able to customize the debounce time to a lower value during the test execution)

> watchpack's watch method takes a third parameter that is the startTime, which is then used to check if the file mtime is >= of the given startTime and not emit the change event if it is not (e.g. see here and here in the watchpack sources), we do set the startTime to Date.now() in how watcher utils.

> watchpak seems to be internally using a FS_ACCURACY variable that may effect the behavior (in other words if the change event will be emitted or not), initially defined here.

Do you know of any good resource where I can read about these in more detail or Should I look into source code more?

> one is the debounce time we do have (changes of making that trigger a failure reduce by making as able to customize the debounce time to a lower value during the test execution)

I am still unsure about how 👆 this affects race condition.

Reply:
> I am still unsure about how :point_up_2: this affects race condition.

@ankushduacodes me too, I really hate when I feel to be missing something going on that would explain an unexpected behavior. Sorry for keeping you waiting a bit, I wanted to give it a try and my only macos machine isn't in a great shape (and wifi not working anymore is one of the most annoying issues it does have :-P)

Anyway I've been able to give it a try to run this test on macos and... not that suprisingly there was one more OS specific behavior that we missed to notice ;-)

To identify what it was happening I opted to add a temporary console.log call inside the `watcher.on('change', filePath => ...);` in `src/watcher.js` and executed the stress test again... well turned out that on macos a `change` event is being fired also for the directory itself from time to time (but pretty often, even on my old macos, I imagine one more recent and in a better shape may trigger it even more often) which is something that wasn't apparently happening on Linux.

To prevent that from happening (we are not interested in that anyway), we can change that `shouldWatchFile: () => true` callback we use in this test and set it to `shouldWatchFile: (filePath) => filePath !== tmpPath`, which will basically filter out that change event and prevent the failures related to this OS-specific difference in behavior.

Besides that I had to add another `await waitDebounce();` right before the `sinon.assert.callCount(onChange, ...)` (as I did mention in one of the previous comments as something that may help to make the test a bit more stable), otherwise the test was still failing pretty soon in the stress test sequence because the onChange callback wasn't called yet by the time we were calling the assertion.

would you mind to try to apply these two small changes and see if it does make the test able to pass the stress test as it did for me? (I kept everything else as is in this version of the PR, `debounceTime` set to `10` and `waitDebounce` using `debounceTime*2`).

if it does make the stress test to pass also on your machine I think that it would be good to update the patch with these two small changes and then land this. how that sounds to you?

Reply:
I think I understand what you mean by the following (Please correct me if I am wrong):
> To prevent that from happening (we are not interested in that anyway), we can change that shouldWatchFile: () => true callback we use in this test and set it to shouldWatchFile: (filePath) => filePath !== tmpPath, which will basically filter out that change event and prevent the failures related to this OS-specific difference in behavior.

We are basically ignoring the `change` event emitted for the directory by returning `false` and basically ignoring the filePath for that event.

Reply:
> We are basically ignoring the change event emitted for the directory by returning false and basically ignoring the filePath for that event from inside of proxyFileChanges() function.

@ankushduacodes yep that's exactly what I mean with that :+1:

> I will try out the tests and will update you on it soon!..

:+1: Thanks a lot for helping me with investigating this!!!

Reply:
@rpl The tests ran successfully this time with the changes you mentioned. 
For me the `change` event on the directory was being emitted every 4 iterations on average.

<br>
Tiny reminder/info request:

>>watchpack's watch method takes a third parameter that is the startTime, which is then used to check if the file mtime is >= of the given startTime and not emit the change event if it is not (e.g. see here and here in the watchpack sources), we do set the startTime to Date.now() in how watcher utils.

>>watchpak seems to be internally using a FS_ACCURACY variable that may effect the behavior (in other words if the change event will be emitted or not), initially defined here.

>Do you know of any good resource where I can read about these in more detail or Should I look into source code more?

Reply:
>> watchpak seems to be internally using a FS_ACCURACY variable that may effect the behavior (in other words if the change event will be emitted or not), initially defined here.
>  Do you know of any good resource where I can read about these in more detail or Should I look into source code more?

@ankushduacodes Was this question specifically about that FS_ACCURACY variable? it is an implementation detail of the watchpack package and so their source code is likely the only place where you can get more details from.

I doubt that it would be documented anywhere else, in some cases to get a bit more context about the actual reasons behind a particular line of code I do use the git blame info to find the pull request that a particular change has been merged from and look into the pull request review comments in case there are any more details useful to get a more complete picture.

Honestly I didn't do it in this case, based on how we use the dependency in web-ext those details don't really matter that much, and so I mainly focused my interest on the details that did matter for keeping the test stable over a big enough number of runs.